pkg/wekafs/dirvolume.go

```diff
@@ -0,0 +1,388 @@
+	glog.V(4).Infoln("Getting existing quota for volume", v.GetId())
+	fs, err := v.getFilesystemObj()
+	if err != nil {
+		return nil, err
+	}
+	inodeId, err := v.getInodeId(mountPath)
+	if err != nil {
+		return nil, err
+	}
+	ret, err := v.apiClient.GetQuotaByFileSystemAndInode(fs, inodeId)
+	if ret != nil {
+		glog.V(4).Infoln("Successfully acquired existing quota for volume", v.GetId(), ret.GetQuotaType(), ret.GetCapacityLimit())
+	}
+	return ret, err
+}
+
+func (v DirVolume) getSizeFromQuota(mountPath string) (uint64, error) {
+	q, err := v.getQuota(mountPath)
+	if err != nil {
+		return 0, err
+	}
+	if q != nil {
+		return q.GetCapacityLimit(), nil
+	}
+	return 0, errors.New("could not fetch quota from API")
+}
+
+func (v DirVolume) getSizeFromXattr(mountPath string) (uint64, error) {
+	if capacityString, err := xattr.Get(v.getFullPath(mountPath), xattrCapacity); err == nil {
+		if capacity, err := strconv.ParseInt(string(capacityString), 10, 64); err == nil {
+			return uint64(capacity), nil
+		}
+		return 0, ErrBadXattrOnVolume
+	}
+	return 0, ErrNoXattrOnVolume
+}
+
+//getSizeFromDf used for GetCapacity, since we don't have secrets to check API
+func (v DirVolume) getSizeFromDf(mountPath string) (uint64, error) {
+	var volStat, parentStat unix.Statfs_t
+	err := unix.Statfs(v.getFullPath(mountPath), &volStat)
+	if err != nil {
+		return 0, err
+	}
+
+	errParent := unix.Statfs(filepath.Dir(v.getFullPath(mountPath)), &parentStat)
+	if errParent != nil {
+		return 0, errParent
+	}
+	volSize := volStat.Blocks * uint64(volStat.Bsize)
+	// Available blocks * size per block = available space in bytes
+	glog.V(3).Infoln("Volume", v.GetId(), "size is", volSize, "bytes")
+	return volSize, nil
+}
+
+func (v DirVolume) getFilesystemObj() (*apiclient.FileSystem, error) {
+	fs, err := v.apiClient.GetFileSystemByName(v.Filesystem)
+	if err != nil {
+		return nil, err
+	}
+	return fs, nil
+}
+
+func (v DirVolume) Mount(mounter *wekaMounter, xattr bool) (string, error, UnmountFunc) {
+	if xattr {
+		return mounter.MountXattr(v.Filesystem)
+	}
+	return mounter.Mount(v.Filesystem)
+}
+
+func (v DirVolume) Unmount(mounter *wekaMounter) error {
+	return mounter.Unmount(v.Filesystem)
+}
+
+func (v DirVolume) Exists(mountPoint string) (bool, error) {
+	if !PathExists(v.getFullPath(mountPoint)) {
+		glog.Infof("Volume %s not found on filesystem %s", v.GetId(), v.Filesystem)
+		return false, nil
+	}
+	if err := pathIsDirectory(v.getFullPath(mountPoint)); err != nil {
+		glog.Errorf("Volume %s is unusable: path %s is a not a directory", v.GetId(), v.Filesystem)
+		return false, status.Error(codes.Internal, err.Error())
+	}
+	glog.Infof("Volume %s exists and accessible via %s", v.id, v.getFullPath(mountPoint))
+	return true, nil
+}
+
+func (v DirVolume) Create(mountPath string, enforceCapacity bool, capacity int64) error {
+	volPath := v.getFullPath(mountPath)
+	if err := os.MkdirAll(volPath, 0750); err != nil {
```

Comment:
We have been asked few times around the default permissions we give to CSI volumes
Should we make it configurable in SC?

Reply:
if it's doable - why not

Reply:
It is too complex. MEanwhile TODO